server/src/main/java/io/crate/execution/engine/collect/PKLookupOperation.java

```diff
@@ -96,29 +96,21 @@ public static Doc lookupDoc(IndexShard shard,
             if (docIdAndVersion == null) {
                 return null;
             }
-            SourceFieldVisitor visitor = new SourceFieldVisitor();
             try {
-                StoredFields storedFields = docIdAndVersion.reader.storedFields();
-                storedFields.document(docIdAndVersion.docId, visitor);
+                StoredRow storedRow
+                    = storedRowLookup.getStoredRow(new ReaderContext(docIdAndVersion.reader.getContext()), docIdAndVersion.docId);
+                return new Doc(
```

Comment:
In the caller method (`getDocument`) `StoredRowLookup` is marked as `@Nullable`. 

Shouldn't we have a null check here?
We can also add `@Nullable` annotation to the `StoredRowLookup` here.

Reply:
I think NULL scenario corresponds to former SourceParser = null, ie 
> // No source is required for simple inserts and duplicate detection

We have tests for duplicate primary keys but I think it hits different code branch:
 (`tryInsertFirst = true` and try `index` right away, which fails later with `VersionConflictException`).
 
I thought we can have null on UPDATE ... ON CONFLICT DO NOTHING, but we don't hit this code if we have `updateAssignments`.

Basically, `StoredRowLookup` can be null only if `tryInsertFirst = false` and there are no `updateAssignments` - is it even a valid scenario?

Reply:
I've added some assertions in TransportShardUpsertAction to make it clearer what's expected here, and removed the `@Nullable` annotation on `getDocument`.